src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexerOutpointsRepository.cs

```diff
@@ -0,0 +1,104 @@
+{
+    /// <summary>Repository for <see cref="OutPointData"/> items with cache layer built in.</summary>
+    public class AddressIndexerOutpointsRepository : MemoryCache<string, OutPointData>
+    {
+        private const string DbOutputsDataKey = "OutputsData";
+
+        /// <remarks>Should be protected by <see cref="LockObject"/></remarks>
+        private readonly LiteCollection<OutPointData> addressIndexerOutPointData;
+
+        private readonly ILogger logger;
+
+        private readonly int maxCacheItems;
+
+        public AddressIndexerOutpointsRepository(LiteDatabase db, ILoggerFactory loggerFactory, int maxItems = 60_000)
+        {
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.addressIndexerOutPointData = db.GetCollection<OutPointData>(DbOutputsDataKey);
+
+            this.maxCacheItems = maxItems;
+        }
+
+        public double GetLoadPercentage()
+        {
+            return Math.Round(this.totalSize / (this.maxCacheItems / 100.0), 2);
+        }
+
+        public void AddOutPointData(OutPointData outPointData)
+        {
+            this.AddOrUpdate(new CacheItem(outPointData.Outpoint, outPointData, 1));
+        }
+
+        public void RemoveOutPointData(OutPoint outPoint)
+        {
+            lock (this.LockObject)
+            {
+                if (this.Cache.TryGetValue(outPoint.ToString(), out LinkedListNode<CacheItem> node))
+                {
+                    this.Cache.Remove(node.Value.Key);
+                    this.Keys.Remove(node);
+                    this.totalSize -= 1;
+                }
+
+                if (!node.Value.Dirty)
+                    this.addressIndexerOutPointData.Delete(outPoint.ToString());
+            }
+        }
+
+        protected override void ItemRemovedLocked(CacheItem item)
+        {
+            base.ItemRemovedLocked(item);
+
+            if (item.Dirty)
+                this.addressIndexerOutPointData.Upsert(item.Value);
+        }
+
+        public bool TryGetOutPointData(OutPoint outPoint, out OutPointData outPointData)
+        {
+            if (this.TryGetValue(outPoint.ToString(), out outPointData))
+            {
+                this.logger.LogTrace("(-)[FOUND_IN_CACHE]:true");
+                return true;
+            }
+
+            // Not found in cache - try find it in database.
+            outPointData = this.addressIndexerOutPointData.FindById(outPoint.ToString());
+
+            if (outPointData != null)
+            {
+                this.AddOutPointData(outPointData);
+                this.logger.LogTrace("(-)[FOUND_IN_DATABASE]:true");
+                return true;
+            }
+
+            return false;
+        }
+
+        public void SaveAllItems()
+        {
+            lock (this.LockObject)
+            {
+                CacheItem[] dirtyItems = this.Keys.Where(x => x.Dirty).ToArray();
+                this.addressIndexerOutPointData.Upsert(dirtyItems.Select(x => x.Value));
```

Comment:
Same here, what if there no items that are dirty?

Reply:
Same answer as above.